tests/components/powerwall/test_switch.py

```diff
@@ -0,0 +1,166 @@
+"""Test for Powerwall off-grid switch."""
+
+from unittest.mock import Mock, patch
+
+from tesla_powerwall import GridStatus
+
+from homeassistant.components.powerwall.const import DOMAIN
+from homeassistant.components.switch import (
+    DOMAIN as SWITCH_DOMAIN,
+    SERVICE_TURN_OFF,
+    SERVICE_TURN_ON,
+)
+from homeassistant.const import ATTR_ENTITY_ID, CONF_IP_ADDRESS, STATE_OFF, STATE_ON
+from homeassistant.helpers import entity_registry as ent_reg
+
+from .mocks import _mock_powerwall_with_fixtures
+
+from tests.common import MockConfigEntry
+
+ENTITY_ID = "switch.take_powerwall_off_grid"
+
+
+async def test_entity_registry(hass):
+    """Test powerwall off-grid switch device."""
+    mock_powerwall = await _mock_powerwall_with_fixtures(hass)
+
+    config_entry = MockConfigEntry(domain=DOMAIN, data={CONF_IP_ADDRESS: "1.2.3.4"})
+    config_entry.add_to_hass(hass)
+    with patch(
+        "homeassistant.components.powerwall.config_flow.Powerwall",
+        return_value=mock_powerwall,
+    ), patch(
```

Comment:
```suggestion
    with patch(
```

https://github.com/home-assistant/core/pull/86357/files#r1083345302

I'd actually make this a fixture so you don't have to patch it every time

Reply:
I'm 99% sure I need to do this in each test, as I'm modifying the value of the mock immediately before in some tests.

If there's a better way to modify a mocked `homeassistant.components.powerwall.Powerwall` instance then I'd very much be happy to switch to fixtures and avoid all the duplication :) (sorry, python newbie here)

Reply:
Note: I've extracted this setup into a method to make it a little clearer / avoid duplication

Reply:
I was thinking a `pytest.fixture` might make sense here.

Reply:
Not sure I follow, sorry.  Can't find any relevant examples in the powerwall component.  Is there an example somewhere I can follow allong with here?

Reply:
What you have is fine.

I was thinking something like this:

```
@pytest.fixture(name="mock_powerwall")
async def mock_powerwall_fixture(hass):
    """Fixture for powerwall."""
    mock_powerwall = await _mock_powerwall_with_fixtures(hass)
    config_entry = MockConfigEntry(domain=DOMAIN, data={CONF_IP_ADDRESS: "1.2.3.4"})
    config_entry.add_to_hass(hass)
    with patch(
        "homeassistant.components.powerwall.Powerwall", return_value=mock_powerwall
    ):
        assert await hass.config_entries.async_setup(config_entry.entry_id)
        await hass.async_block_till_done()    
        yield mock_powerwall
```

```
async def test_turn_off_without_entity_id(hass, mock_powerwall):
    """Test switch turn off all switches."""
    mock_powerwall.get_grid_status = Mock(return_value=GridStatus.CONNECTED)

    await hass.services.async_call(
        SWITCH_DOMAIN, SERVICE_TURN_OFF, {ATTR_ENTITY_ID: "all"}, blocking=True
    )

    state = hass.states.get(ENTITY_ID)
    assert state.state == STATE_OFF
```

Reply:
Ahh great, I reckon I can sort that out.  Thanks again for the help

Reply:
Done!